modules/core/shared/src/main/scala-3/io/circe/derivation/ConfiguredDecoder.scala

```diff
@@ -19,10 +20,22 @@ trait ConfiguredDecoder[A](using conf: Configuration) extends Decoder[A]:
 
   /** Decodes a class/object/case of a Sum type handling discriminator and strict decoding. */
   private def decodeSumElement[R](c: HCursor)(fail: DecodingFailure => R, decode: Decoder[A] => ACursor => R): R =
+
     def fromName(sumTypeName: String, cursor: ACursor): R =
-      constructorNames.indexOf(sumTypeName) match
-        case -1 => fail(DecodingFailure(s"type $name has no class/object/case named '$sumTypeName'.", cursor.history))
-        case index => decode(elemDecoders(index).asInstanceOf[Decoder[A]])(cursor)
+      def findDecoder(decoder: Decoder[?], searched: List[ConfiguredDecoder[?]]): Option[Decoder[?]] =
+        decoder match {
+          case cd: ConfiguredDecoder[?] with SumOrProduct if !searched.contains(cd) && cd.isSum =>
```

Comment:
I feel like its probably faster to check if `cd` is a sum type first and know if you need to bother checking the list.
```suggestion
          case cd: ConfiguredDecoder[?] with SumOrProduct if cd.isSum && !searched.contains(cd) =>
```